src/util/flatRender.js

```diff
@@ -0,0 +1,193 @@
+import {
+  contextOf,
+  equalPath,
+  getThoughtsRanked,
+  head,
+  isDescendant,
+  isFunction,
+  pathToContext,
+  unroot,
+} from '../util'
+
+import { RANKED_ROOT } from '../constants'
+
+const MAX_DEPTH_FROM_CURSOR = 7
+
+// recursively finds all the visible thought and returns a flat array
+const getFlatArray = ({
+  startingPath,
+  cursor,
+  children,
+  isLeaf,
+  showHiddenThoughts,
+  isParentCursorAncestor = true,
+  isCursorChildren = false,
+} = {}) => {
+  const parentNode = head(startingPath) || RANKED_ROOT[0]
+
+  const subThoughts = children || getThoughtsRanked(startingPath)
+  const checkIfContextOfCursor = equalPath(startingPath, contextOf(cursor))
+
+  // iterate subthoughts
+  return subThoughts.reduce((acc, child) => {
+    const childPath = unroot(startingPath.concat(child))
+    const value = child.value
+
+    // isParentCursorAncestor is used to prevent calling isDescendant everytime
+    // if the parent thought is already not an ancestor of the cursor then we don't need to call it everytime for its descendants
+    const isCursorAncestor =
+      isParentCursorAncestor &&
+      isDescendant(pathToContext(childPath), pathToContext(cursor))
+    const isCursor = equalPath(cursor, childPath)
+    const childPathLength = childPath.length
+
+    const children = getThoughtsRanked(childPath)
```

Comment:
I have a specific reason for this. We need to calculate total number of visible children if a thoughts expands to show either `▸` or  `•`. So to do that we need to calculate no of hidden nodes (all meta children + all children that has attribute `=hidden`).

Since we already calculate children of a node, we can filter the meta children at a single iteration of `children` array. But we also need to find if each children has hidden attribute or not. To do that we again have to iterate children of each child in that array and check if they have at least one meta `=hidden` thought. That means nested iteration. So at every node it will iterate over its children and all of its grand children.

We know that we will eventually iterate over grandchildren deeper into recursion then we shouldn't iterate it at the parent level. So every node will calculate it's children and know if its hidden or not. We can count it and return it to the parent node. 

Due to recursion logic it is a bit complicated to convey my idea. If you feel this is unnecessary we can remove `depthInfo` logic.

Reply:
Okay, I understand better now. The recursion is depth-first, so you are using the recursion to "pass" the number of hidden thoughts from child to parent.

> We need to calculate total number of visible children if a thoughts expands to show either `▸` or `•`. So to do that we need to calculate no of hidden nodes (all meta children + all children that has attribute `=hidden`).

Technically we just need a boolean that is `true` if there is at least one visible child, not the total number of them.

> Since we already calculate children of a node, we can filter the meta children at a single iteration of `children` array. But we also need to find if each children has hidden attribute or not. To do that we again have to iterate children of each child in that array and check if they have at least one meta `=hidden` thought. That means nested iteration. So at every node it will iterate over its children and all of its grand children.

Good point. We want to avoid that nested iteration.

> We know that we will eventually iterate over grandchildren deeper into recursion then we shouldn't iterate it at the parent level. So every node will calculate it's children and know if its hidden or not. We can count it and return it to the parent node.

Yes, although if all we need is a boolean, we get rid of the counting logic.

> Due to recursion logic it is a bit complicated to convey my idea. If you feel this is unnecessary we can remove `depthInfo` logic.

I don't like the added complexity, but it's probably worth it to avoid the nested iteration.

Reply:
All of that aside, I still have my original question: Is there any reason we can't filter out hidden children here (assuming `showHiddenThoughts` is false)? Is there anything that actually needs to know about hidden children in this algorithm? As far as I can tell, we just need to know that there is *at least one visible child*, not how many hidden children there are.

Reply:
> All of that aside, I still have my original question: Is there any reason we can't filter out hidden children here (assuming `showHiddenThoughts` is false)? Is there anything that actually needs to know about hidden children in this algorithm? As far as I can tell, we just need to know that there is _at least one visible child_, not how many hidden children there are.

Yes you are right. We can filter all the meta thoughts here . We also can filter out hidden thoughts with nested iteration. We can filter all meta thoughts and depend on `getFlatArray` to return if there are any visible nodes available.